### PR TITLE
[WIP] Add ZeroShotObjectDetectionPipeline (#18445)

### PR DESCRIPTION
**Pull request by sahamrit**
# What does this PR do?

This PR adds the `ZeroShotObjectDetectionPipeline`. It is tested on `OwlViTForObjectDetection` model and should enable the inference following inference API

```
from transformers import pipeline

pipe = pipeline("zero-shot-object-detection")
pipe("cats.png", ["cat", "remote"])
```

This pipeline could default to the [https://huggingface.co/google/owlvit-base-patch32](https://huggingface.co/google/owlvit-base-patch32) checkpoint

Fixes # (`18445`)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Link to the [Issue](https://github.com/huggingface/transformers/issues/18445)
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@alaradirik @Narsil 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/auto/modeling_auto.py
Comment (sahamrit) [inline src/transformers/models/auto/modeling_auto.py:449]: This maybe the cause of [https://app.circleci.com/pipelines/github/huggingface/transformers/46856/workflows/0831cfcd-9a88-4420-a868-f131b36754a8/jobs/552799?invite=true#step-111-5713](test failure).

Any suggestions ? Does creating a new  `MODEL_FOR_XX` mapping help?

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (sahamrit) [inline tests/pipelines/test_pipelines_zero_shot_object_detection.py:97]: The reason of skipping this test is `hf-internal-testing/tiny-random-owlvit` throws the following error.

Code to reproduce bug:
```
from transformers import (
    MODEL_FOR_OBJECT_DETECTION_MAPPING,
    AutoFeatureExtractor,
    AutoTokenizer,
    AutoModelForObjectDetection,
    ZeroShotObjectDetectionPipeline,
    is_vision_available,
    pipeline,
)
from transformers.testing_utils import (
    is_pipeline_test,
    nested_simplify,
    require_tf,
    require_torch,
    require_vision,
    slow,
)

from PIL import Image

model_id = "hf-internal-testing/tiny-random-owlvit"

model = AutoModelForObjectDetection.from_pretrained(model_id)
feature_extractor = AutoFeatureExtractor.from_pretrained(model_id)
tokenizer = AutoTokenizer.from_pretrained(model_id)

object_detector = ZeroShotObjectDetectionPipeline(
    model=model, feature_extractor=feature_extractor, tokenizer=tokenizer
)

outputs = object_detector(
    "./tests/fixtures/tests_samples/COCO/000000039769.png",
    candidate_labels=["cat", "remote", "couch"],
    threshold=0.0,
)
```

Traceback along with some warning

```
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Some weights of OwlViTForObjectDetection were not initialized from the model checkpoint at hf-internal-testing/tiny-random-owlvit and are newly initialized: ['class_head.logit_scale.bias', 'class_head.dense0.weight', 'box_head.dense0.weight', 'box_head.dense1.weight', 'layer_norm.bias', 'class_head.dense0.bias', 'class_head.logit_shift.bias', 'box_head.dense1.bias', 'class_head.logit_scale.weight', 'box_head.dense2.weight', 'layer_norm.weight', 'box_head.dense0.bias', 'box_head.dense2.bias', 'class_head.logit_shift.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Traceback (most recent call last):
  File "/home/azureuser/cloudfiles/code/Users/asahu.cool/Work/Repos/transformers/test.py", line 34, in <module>
    threshold=0.0,
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/pipelines/zero_shot_object_detection.py", line 64, in __call__
    return super().__call__(images, **kwargs)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/pipelines/base.py", line 1074, in __call__
    return self.run_single(inputs, preprocess_params, forward_params, postprocess_params)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/pipelines/base.py", line 1081, in run_single
    model_outputs = self.forward(model_inputs, **forward_params)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/pipelines/base.py", line 990, in forward
    model_outputs = self._forward(model_inputs, **forward_params)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/pipelines/zero_shot_object_detection.py", line 94, in _forward
    outputs = self.model(**model_inputs)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1384, in forward
    (pred_logits, class_embeds) = self.class_predictor(image_feats, query_embeds, query_mask)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1256, in class_predictor
    (pred_logits, image_class_embeds) = self.class_head(image_feats, query_embeds, query_mask)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1149, in forward
    pred_logits = torch.einsum("...pd,...qd->...pq", image_class_embeds, query_embeds)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/functional.py", line 360, in einsum
    return _VF.einsum(equation, operands)  # type: ignore[attr-defined]
RuntimeError: einsum(): operands do not broadcast with remapped shapes [original->remapped]: [1, 576, 8]->[1, 576, 1, 8] [1, 3, 16]->[1, 1, 3, 16]
```

**Review by alaradirik — COMMENTED** [review]
Thank you for your contribution and sorry for the delayed review! We are really stoked to see the addition of this pipeline and planning to add new models to it in the near future.

I left a few comments to address the issues but to summarize:
- Although OWL-ViT uses CLIP as its base model, the output and preprocessing steps are different
- The tests of zero-shot-image-classification are not applicable to OWL-ViT as the outputs are different

These should be easy to fix and I'm looking forward to reviewing this PR again :)

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (sahamrit) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:97]: Hi @alaradirik . [hf-internal-testing/tiny-random-owlvit](https://huggingface.co/hf-internal-testing/tiny-random-owlvit/tree/main) is present on hub

File: src/transformers/pipelines/zero_shot_object_detection.py
Comment (sahamrit) [reply src/transformers/pipelines/zero_shot_object_detection.py:62]: Hi, the output of __call__ function should return the post processed output right? Post Processing as show in [OwlViTProcessor.post_process](https://huggingface.co/docs/transformers/main/en/model_doc/owlvit#transformers.OwlViTProcessor.post_process) and [Post-processing model predictions
](https://github.com/huggingface/notebooks/blob/main/examples/zeroshot_object_detection_with_owlvit.ipynb) do return label, score and box

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (alaradirik) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:97]: My mistake, sorry about that! 

The error is most likely due to calling feature extractor and tokenizer separately instead of using OwlViTProcessor, which makes sure that the `input_ids` has the correct shape.

File: src/transformers/pipelines/zero_shot_object_detection.py
Comment (sahamrit) [reply src/transformers/pipelines/zero_shot_object_detection.py:62]: Hi can you please take a look above?

File: src/transformers/pipelines/zero_shot_object_detection.py
Comment (thedch) [inline src/transformers/pipelines/zero_shot_object_detection.py:105]: can we avoid the hardcoded `[0]` and use `more_itertools.one` instead? More explicit

File: src/transformers/pipelines/zero_shot_object_detection.py
Comment (sahamrit) [reply src/transformers/pipelines/zero_shot_object_detection.py:86]: Hi @alaradirik , I guess in __init__.py of pipelines we don't auto set the processor, like we do for [feature extractor](https://github.com/huggingface/transformers/blob/main/src/transformers/pipelines/__init__.py#:~:text=kwargs%5B%22feature_extractor%22%5D) & tokenizer. Do you want to add an explicit check if model_name defaults to OwlVit then use OwlVitProcessor ?

Comment (alaradirik) [reply src/transformers/pipelines/zero_shot_object_detection.py:86]: Hi @sahamrit, yes that is true. Doing an explicit check might cause test failures due to unexpected outputs but we can either replicate the behaviour of `OwlViTProcessor` using the tokenizer and feature_extractor or add a processor attribute to the Pipeline class. Replicating the preprocessing steps within `pipelines/zero_shot_object_detection.py` would be the easiest and preferable as the XXXProcessor classes have no clearly defined API yet.

You can find the relevant preprocessing steps to pad and stack the tokenized text queries on lines 82-122 of the [processing_owlvit.py](https://github.com/huggingface/transformers/blob/main/src/transformers/models/owlvit/processing_owlvit.py) file.

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (sahamrit) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:97]: Hi @alaradirik , it seems `hf-internal-testing/tiny-random-owlvit` isn't even working with the demo notebook code.

**Code to reproduce bug**
`from transformers import OwlViTProcessor, OwlViTForObjectDetection
from PIL import Image
import os
import skimage
import numpy as np
from PIL import Image
import torch


image = skimage.data.astronaut()
image = Image.fromarray(np.uint8(image)).convert("RGB")
text_queries = ["human face", "rocket", "nasa badge", "star-spangled banner"]
model_id = "hf-internal-testing/tiny-random-owlvit"

if torch.cuda.is_available():
    device = torch.device("cuda")
else:
    device = torch.device("cpu")

model = OwlViTForObjectDetection.from_pretrained(model_id)
processor = OwlViTProcessor.from_pretrained(model_id)

inputs = processor(text=text_queries, images=image, return_tensors="pt").to(device)

model = model.to(device)
model.eval()

with torch.no_grad():
    outputs = model(**inputs)`
    
**Error**
`WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Some weights of OwlViTForObjectDetection were not initialized from the model checkpoint at hf-internal-testing/tiny-random-owlvit and are newly initialized: ['box_head.dense0.bias', 'box_head.dense2.bias', 'box_head.dense1.weight', 'box_head.dense1.bias', 'class_head.logit_shift.bias', 'layer_norm.bias', 'layer_norm.weight', 'class_head.logit_scale.bias', 'class_head.logit_scale.weight', 'class_head.dense0.weight', 'box_head.dense2.weight', 'class_head.dense0.bias', 'box_head.dense0.weight', 'class_head.logit_shift.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Traceback (most recent call last):
  File "/home/azureuser/cloudfiles/code/Users/asahu.cool/Work/Repos/transformers/test.py", line 46, in <module>
    outputs = model(**inputs)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1384, in forward
    (pred_logits, class_embeds) = self.class_predictor(image_feats, query_embeds, query_mask)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1256, in class_predictor
    (pred_logits, image_class_embeds) = self.class_head(image_feats, query_embeds, query_mask)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1149, in forward
    pred_logits = torch.einsum("...pd,...qd->...pq", image_class_embeds, query_embeds)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/functional.py", line 360, in einsum
    return _VF.einsum(equation, operands)  # type: ignore[attr-defined]
RuntimeError: einsum(): operands do not broadcast with remapped shapes [original->remapped]: [1, 576, 8]->[1, 576, 1, 8] [1, 4, 16]->[1, 1, 4, 16]`


Seeing the error it seems the query and image embeddings are having different dimensions - 16 vs 8 respectively

Comment (alaradirik) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:97]: Hi @sahamrit, thanks for bringing this up. @lewtun could you take a look at `hf-internal-testing/tiny-random-owlvit`?

Comment (sahamrit) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:97]: Hi @alaradirik , @lewtun . Any updates on this? I wanted to close this PR by this week since am on vacation next week.

Comment (lewtun) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:97]: Hey @sahamrit I see the problem is that the test model `hf-internal-testing/tiny-random-owlvit` is missing the object detection head. 

We're currently overhauling these tiny models - @ydshieh would the best course of action be for me to create a `hf-internal-testing/tiny-random-owlvit-object-detection` model for just this task?

Comment (ydshieh) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:97]: @lewtun Yes!

Comment (lewtun) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:97]: OK great, I'll add this right away and report back when it's up :)

Comment (lewtun) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:97]: Sorry for the delay - had some issues shrinking the model correctly. 

You can use this ID for the tests: https://huggingface.co/hf-internal-testing/tiny-random-owlvit-object-detection

**Comment by ydshieh** [comment]
Hi, just seeing the merge messed up the commit history. There are 377 changes, which is impossible for the review and merge to `main`.

I suggest to reset to the last clean commit locally. Then use  `git rebase main` to keep update with `main` (after pulling the latest changes from remote `main` into local `main`).

**Comment by sahamrit** [comment]
> Hi, just seeing the merge messed up the commit history. There are 377 changes, which is impossible for the review and merge the PR into `main`.
> 
> I suggest to reset to the last clean commit locally. Then use `git rebase main` to keep update with `main` (after pulling the latest changes from remote `main` into local `main`). Or any way works (as I am not sure what causes the current git status)

Hi @ydshieh sorry for that. Was in a hurry to wrap the PR since I was going for vacation. Messed up in rebasing. Have reverted to stable commit. Will add the correct changes once I am back!

**Comment by ydshieh** [comment]
No problem, @sahamrit! I am super happy that you are able to get back to the stable commit 💯 . Have a nice vacation!

**Review by alaradirik — COMMENTED** [review]
Hi there, thank you once again for the contribution!

The PR is almost ready to merge, I just left a few comments in zero_shot_object_detection.py to ensure preprocess and forward methods support batch processing and suggested an optional format change in post_process to similarly support batch processing. You can directly apply the suggested changes.

I saw that code quality check and tests were not passing, you can always check if your code is correctly formatted and passes the tests with:
```
make fixup
pytest tests/pipelines/test_pipelines_zero_shot_object_detection.py
```

This is the code I use to test the pipeline:
```
import skimage
import numpy as np
from PIL import Image
from transformers import pipeline

image = skimage.data.astronaut()
image = Image.fromarray(np.uint8(image)).convert("RGB")
text_queries = [["human face", "rocket", "nasa badge", "star-spangled banner"], ["a person"]]

pipe = pipeline("zero-shot-object-detection")
inputs = {"images": [image, image], "text_queries": text_queries}

preprocessed = pipe.preprocess(inputs=inputs)
model_outputs = pipe.forward(preprocessed)
outputs = pipe.postprocess(model_outputs)
```

**Comment by sahamrit** [comment]
Hi @alaradirik , can you review the changes?

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (alaradirik) [inline tests/pipelines/test_pipelines_zero_shot_object_detection.py:64]: This test fails as threshold is not included in `ZeroShotObjectDetectionPipeline` arguments. Could you add the `threshold` argument to __call__ and make the necessary changes to preprocess, _forward and postprocess methods?

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (alaradirik) [inline tests/pipelines/test_pipelines_zero_shot_object_detection.py:96]: Same comment here, threshold is not included in the call signature yet.

**Review by alaradirik — COMMENTED** [review]
Hi again, everything looks good but all slow tests are failing as the threshold argument is not included in the call signature of ZeroShotObjectDetectionPipeline. Could you either remove the threshold argument from the tests or add the threshold argument to the pipeline such that the pipeline could be called like this:

```
pipe = pipeline("zero-shot-object-detection")
outputs = pipe(images=[image, image], text_queries=text_queries, threshold=threshold)
```

or like this:
```
inputs = {"images": [image, image], "text_queries": text_queries, "threshold": threshold}
preprocessed = pipe.preprocess(inputs=inputs)
model_outputs = pipe.forward(preprocessed)
outputs = pipe.postprocess(model_outputs)
```

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (sahamrit) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:96]: Hi @alaradirik , all the slow tests are passing for me can you please check.

```
[gw4] PASSED tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_large_model_pt 
tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_large_model_tf 
[gw4] SKIPPED tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_large_model_tf 
tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_pt_OwlViTConfig_OwlViTForObjectDetection_CLIPTokenizerFast_OwlViTFeatureExtractor 
[gw4] PASSED tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_pt_OwlViTConfig_OwlViTForObjectDetection_CLIPTokenizerFast_OwlViTFeatureExtractor 
tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_pt_OwlViTConfig_OwlViTForObjectDetection_CLIPTokenizer_OwlViTFeatureExtractor 
[gw4] PASSED tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_pt_OwlViTConfig_OwlViTForObjectDetection_CLIPTokenizer_OwlViTFeatureExtractor 
tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_small_model_pt 
[gw4] PASSED tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_small_model_pt 
tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_small_model_tf 
[gw4] SKIPPED tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_small_model_tf 
tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_threshold No model was supplied, defaulted to google/owlvit-base-patch32 and revision 17740e1 (https://huggingface.co/google/owlvit-base-patch32).
Using a pipeline without specifying a model name and revision in production is not recommended.

[gw4] PASSED tests/pipelines/test_pipelines_zero_shot_object_detection.py::ZeroShotObjectDetectionPipelineTests::test_threshold
```

File: src/transformers/pipelines/zero_shot_object_detection.py
Comment (sahamrit) [inline src/transformers/pipelines/zero_shot_object_detection.py:106]: The threshold param flows in via kwargs. You can also refer [zero_shot_image_classification pipeline](https://github.com/huggingface/transformers/blob/main/src/transformers/pipelines/zero_shot_image_classification.py#:~:text=def%20_sanitize_parameters(,return%20preprocess_params%2C%20%7B%7D%2C%20%7B%7D)

File: src/transformers/pipelines/zero_shot_object_detection.py
Comment (sahamrit) [inline src/transformers/pipelines/zero_shot_object_detection.py:70]: I forgot to add the threshold param in the __call__ documentation. I will add this in the final commit along with any other changes if needed

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (alaradirik) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:96]: Sorry, turns out I hadn't rebuilt the library after pulling the changes, all tests are now passing! It's be sufficient to just add the threshold argument to the call signature

File: src/transformers/pipelines/zero_shot_object_detection.py
Comment (alaradirik) [reply src/transformers/pipelines/zero_shot_object_detection.py:70]: @sahamrit could you also rebase your branch to huggingface:main? There are some tests unrelated to your PR failing, rebasing should fix them too.

**Review by alaradirik — APPROVED** [review]
Everything looks good to me! Only thing left is to add the `threshold` argument to the call signature.

I'm adding @LysandreJik  for the final review and approval for the merge.

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (sgugger) [inline tests/pipelines/test_pipelines_zero_shot_object_detection.py:64]:
```suggestion
        batch_outputs = object_detector(examples, threshold=0.0)
```
Fits in one line.

**Review by sgugger — APPROVED** [review]
Thanks a lot for the PR! LGTM but since it's a pipeline, I'll defer to @Narsil before merging :-)

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (Narsil) [inline tests/pipelines/test_pipelines_zero_shot_object_detection.py:103]: I feel like these outputs are unnatural.

Why not `[{"score": 0.67, "label": "remote", "bbox": ...}, {...}, ]` instead ?
The problem with this output is that most users will probably have to zip anyway, and there's room for error if the list sizes don't match.

Comment (Narsil) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:103]: Also why don't we sort by descending score ?

And instead of `threshold` which is likely to be model dependant, why don't we use `top_k` instead ?

(We can keep threshold, but `top_k`  feels more usable to me).

**Review by Narsil — COMMENTED** [review]
Thank you for this PR.

- I suggest to modify the output of the pipeline to be more "natural". (see relevant comment).
- `text_queries` should be renamed `candidate_labels` to be in line with `zero-shot-classification`.

Another note, is that this code is trying too hard to normalize inputs and the code is more bloated than it should IMHO.

Normalization in `__call__` : Should be about making things easier for python callers and producing a UNIQUE output format.

`pipe(image = 'https://...', text_queries=["a", "b", "c'])`
This would imply `[{"image": "https://..", 'candidate_label': 'a'}, {"image": ..., "candidate_label": "b'}, ...]` 

Don't try to support TOO many call formats, otherwise some call formats will be undecidable on what the user actually intents (so don't overdo the list and list of lists basically). 
For instance a list of text could be many texts, or a pretokenizer string (it is used in other pipelines). For this pipeline it does not make sense, but just to give an idea of how intent could be hard to understand.

File: src/transformers/pipelines/zero_shot_object_detection.py
Comment (Narsil) [inline src/transformers/pipelines/zero_shot_object_detection.py:234]: `jax` is actually not supported (yet and unlikely to be added anytime soon) in pipelines. So we can remove that.

File: src/transformers/pipelines/zero_shot_object_detection.py
Comment (Narsil) [inline src/transformers/pipelines/zero_shot_object_detection.py:225]: You shouldn't have to do that.

Padding is not up to the preprocessing.

The better option is to inherit `ChunkPipeline` instead of `Pipeline`. Check out `ZeroShotClassification` if you want to see a little more how it works.

However it's slightly more involved, but it *will* make this code disappear which is better IMO.

**Review by Narsil — COMMENTED** [review]
And thanks for those tests, they are perfect !

**Comment by alaradirik** [comment]
> Thank you for this PR.
> 
> * I suggest to modify the output of the pipeline to be more "natural". (see relevant comment).
> * `text_queries` should be renamed `candidate_labels` to be in line with `zero-shot-classification`.

Hey @Narsil! I suggested using `text_queries` instead because it is a multi-modal model where users query images with free-form text. The queried object is either found or not and the found object's label is not chosen from a selection of candidate labels, so I think it'd make more sense to keep as it is.

**Comment by sahamrit** [comment]
Hi @Narsil , Sure the output labels are taken **exactly** from the input text_queries. The reason of naming it "text_queries" instead of "candidate_labels" in case of zero-shot-image-classification is that, in zero-shot-image-classification pipeline, the [candidate labels are wrapped by the hypothesis template](https://github.com/huggingface/transformers/blob/main/src/transformers/pipelines/zero_shot_image_classification.py#:~:text=candidate_labels%20(%60List%5Bstr,logits_per_image), whereas here the text_queries are free text queries!

Hope it clarifies

**Comment by alaradirik** [comment]
> Are you sure ? I just tried your code, and it seems all the labels stem from the text being sent. Meaning I think there is a 1-1 correspondance between `label` and `text_queries` (meaning `candidate_labels` would be a fine name).
> 

Yes, there is a 1-1 correspondence but I meant only the query text / a single label is evaluated for each object, whereas the label is selected from among multiple candidate labels for `zero-shot-classification`.

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py
Comment (sahamrit) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:103]: I slightly have a different view here. Consider this text_query = ["human-face", "car"] and the image is a street setting. If cars are pretty easily detectable, they will have high scores and dominate the top_k detections. Human Faces in streets being harder to detect will always have a low confidence. Hence playing around with the threshold would be easier as compared to top_k.

@alaradirik  any thoughts?

**Comment by Narsil** [comment]
> Yes, there is a 1-1 correspondence but I meant only the query text / a single label is evaluated for each object, whereas the label is selected from among multiple candidate labels for zero-shot-classification.

I still think that `zero-shot` -> `candidate_labels` logic works. If we reuse names, it means that it's easier on users to discover and use pipelines. The fact that they are slightly different doesn't justify in my eyes the use of a different name.
I would even argue that they are exactly the same and the difference in how they are used are cause by `classification` vs `object-detection` not by what `candidate_labels` are.

**Comment by sahamrit** [comment]
HI @Narsil @alaradirik, kindly review the changes

**Review by alaradirik — APPROVED** [review]
LGTM!

**Review by Narsil — APPROVED** [review]
This looks good to me.

I still think `text_queries` should be renamed `candidate_labels`.
But this can be done later (I can take care of it if you're ok)..

Thank you so much for your contribution, sorry to be late and picky btw !

**Pull request merged**
